sn_client/src/files/upload.rs

```diff
@@ -426,16 +418,158 @@ impl FilesUpload {
+                .await
+            {
+                Ok(cost) => {
+                    debug!("Storecosts retrieved for {:?} {cost:?}", chunk_info.name);
+                    TaskResult::GetStoreCostOK((chunk_info, cost))
+                }
+                Err(err) => {
+                    error!(
+                        "Encountered error {err:?} when getting store_cost of {:?}",
+                        chunk_info.name
+                    );
+                    TaskResult::ErrorEntries(vec![chunk_info])
+                }
+            };
+
+            let _ = get_store_cost_sender.send(cost).await;
+        });
+    }
+
+    fn spawn_paying_thread(
+        &self,
+        mut paying_work_receiver: mpsc::Receiver<Option<(ChunkInfo, PayeeQuote)>>,
+        pay_for_chunk_sender: mpsc::Sender<TaskResult>,
+        batch_size: usize,
+    ) {
+        let files_api = self.api.clone();
+        let verify_store = self.verify_store;
+        let _handle = tokio::spawn(async move {
+            let mut wallet_client = match files_api.wallet() {
+                Ok(wallet) => wallet,
+                Err(err) => {
+                    error!("Failed to open wallet when handling {err:?}");
+                    let _ = pay_for_chunk_sender
+                        .send(TaskResult::ErrorEntries(vec![]))
+                        .await;
+                    return;
+                }
+            };
+            let mut cost_map = BTreeMap::new();
+            let mut chunk_info_map = vec![];
+
+            while let Some(payment) = paying_work_receiver.recv().await {
+                let make_payments = if let Some((chunk_info, quote)) = payment {
+                    let _ = cost_map.insert(chunk_info.name, (quote.1, quote.2));
+                    chunk_info_map.push((chunk_info, quote.0));
+                    cost_map.len() >= batch_size
+                } else {
+                    // using None to indicate as all paid.
+                    !cost_map.is_empty()
+                };
+
+                if make_payments {
+                    let result = match wallet_client.pay_for_records(&cost_map, verify_store).await
+                    {
+                        Ok((storage_cost, royalty_fees)) => {
+                            trace!("Made payments for {} chunks", cost_map.len());
+                            let reply_list = std::mem::take(&mut chunk_info_map);
+                            TaskResult::MakePaymentsOK((
+                                (storage_cost, royalty_fees, wallet_client.balance()),
+                                reply_list,
+                            ))
+                        }
+                        Err(err) => {
+                            let reply_list: Vec<ChunkInfo> = std::mem::take(&mut chunk_info_map)
+                                .iter()
+                                .map(|(chunk_info, _)| chunk_info)
+                                .cloned()
+                                .collect();
```

Comment:
Tiny one to avoid a clone.
```suggestion
                            let reply_list: Vec<ChunkInfo> = std::mem::take(&mut chunk_info_map)
                                .into_iter()
                                .map(|(chunk_info, _)| chunk_info)
                                .collect();
```